### ISSUE TITLE
Crackling Unclean Audio in 1080i5994

### ISSUE THREAD

**Issue by ronag** [CasparCG/Server#387]
Running a channel in 1080i5994 results in crackling audio.

2.0.7

**Comment by HellGore** [comment]
I tried with system-audio and with a Decklink 4K Extreme 6G but I can't reproduce this. Could you provide more information about your system setup and perhaps a test-file you have used when you observed this problem?

**Comment by ronag** [comment]
Decklink Studio

https://www.wetransfer.com/downloads/b15380d019af8f56ef4aa11bd44d2a8a20151026221421/81f897f2e27398e4c08e3887b99676dc20151026221421/58ada9

**Comment by HellGore** [comment]
I have downloaded the file now, so you can remove it if keeping the file public is a problem

**Comment by HellGore** [comment]
Its actually interesting, I get the same audio pops when running the file on a 1080i5000 channel, the audio is the same (video of course out of sync). When looking at the audio meter in the client (fed via OSC) I can see that the pops seems to be when the volume goes up to 0dbf, so it seems to be an audio amplitude issue rather than an audio cadence issue. I'm guessing something with the integer -> float -> integer conversions in the audio_mixer. Probably what we are seeing is an overflow in the calculation so that a sample becomes negative instead of positive or vice versa.

**Comment by HellGore** [comment]
I can verify if I do a MIXER 1 VOLUME 0.9 I don't get the pops, so it is something with the audio mixer

**Comment by ronag** [comment]
Yes

This looks a little suspicious:

`return caspar::array<int32_t>(result.data(), result.size(), true, std::move(result_owner));`

**Comment by HellGore** [comment]
Yes I changed it so that caspar::array is used for audio in 2.1.0 (already used for image data) to be able to have different type erased data owners (it is actually an av_frame in cases when the pan-filter has been used) to avoid unnecessary copying.

However this is not in 2.0.7 so the problem is probably something else in audio_mixer.cpp (same in 2.0 and 2.1, both gets the audio pop)

**Comment by ronag** [comment]
Aah, sorry, I still get confused about which branch is the default one.

**Comment by ronag** [comment]
Yea, I have no idea what is going on with the audio code. Is a bit over my head. Think I'll go with the VOLUME workaround for now.

**Comment by deedos** [comment]
This might be related https://github.com/CasparCG/Server/issues/326

2015-10-27 11:25 GMT-02:00 Robert Nagy <notifications@github.com>:

> Yea, I have no idea what is going on with the audio code. Is a bit over my
> head. Think I'll go with the VOLUME workaround for now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CasparCG/Server/issues/387#issuecomment-151493121>.
>



-- 
Daniel Roviriego
(21) 972361654

**Comment by ronag** [comment]
Another strange thing is the audio cadences for 2994 and 5994 don't really match:

`boost::assign::list_of(1602)(1601)(1602)(1601)(1602)`

vs

`boost::assign::list_of(801) (800) (801) (801) (801)`

Notice the 4th element. 1601 vs 801. I would think that 801 should be 800...

**Comment by HellGore** [comment]
The problem is that `std::numeric_limits<int32_t>::max()` is 2147483647 but `static_cast<float>(2147483647)` results in `2147483648.0f` (+1 from the original) which when casted back to `int32_t` overflows to `-2147483648`.

So we have to think of a way of "clipping" on the right side of the 0 line instead of overflowing via static_cast

**Comment by HellGore** [comment]
First static_casting to int64_t and then truncating via std::min and std::max and then static_casting back to int32_t would work but I don't know about the performance hit when doing this for each sample.

**Comment by ronag** [comment]
The performance should be negligable.

**Comment by HellGore** [comment]
OK, good.

Btw, do you remember if there is there a reason why we are mixing with `float` instead of using `int32_t` all the way inside the audio_mixer?

**Comment by ronag** [comment]
Hm, no I don't remember. Should be fine to use int32_t.

**Comment by ronag** [comment]
Ah, I think it was because of the overflow issue I used float when e.g. combining multiple audio streams. However, I probably didn't consider this specific issue with the signed integer overflows in the conversion back.

**Comment by ronag** [comment]
Ok. From my test this issue does not seem to be related to the audio levels. I just halved the level and it's the same problem.

**Comment by ronag** [comment]
It's a cadence issue. The audio mixer warns for appended zero samples.

**Comment by ronag** [comment]
It's a bug in the audio mixer with interlacing + NTSC audio_cadence. In audio mixer add:

```cpp
BOOST_FOREACH(auto& item, items_)
{
    if (next_audio_streams.find(item.tag) != next_audio_streams.end())
       continue;
  // ...
```

**Comment by HellGore** [comment]
I think you have fixed the cadence issue in 2.1 actually.

Btw, I guess the only expected case of a cadence error occurring should be in the beginning of an audio_stream right? Perhaps we should be prepending the silence instead of appending it. What do you think?

**Comment by ronag** [comment]
No, that won't work since a video clip could start any time in relations to the channel. Which is why in frame_muxer is slightly shifted so the larger values are in the beginning as to avoid underflows.
